File: src/main/java/io/felipeandrade/gsw2/item/tool/GSWHoe.java
package io.felipeandrade.gsw2.item.tool;

import net.fabricmc.fabric.api.client.rendering.v1.ColorProviderRegistry;
import net.minecraft.item.HoeItem;
import net.minecraft.item.ItemStack;
import net.minecraft.item.ToolMaterial;
import net.minecraft.util.Identifier;
import net.minecraft.util.registry.Registry;

import io.felipeandrade.gsw2.material.GSWItemColorProvider;
import io.felipeandrade.gsw2.material.GSWMaterial;

import static io.felipeandrade.gsw2.GSW2Mod.MOD_ID;

public class GSWHoe extends HoeItem implements GSWTool, GSWItemColorProvider {

    protected final String unlocalizedName;
    private final GSWMaterial material;

    public GSWHoe(String unlocalizedName, GSWMaterial material, ToolMaterial toolMaterial, float attackSpeed, Settings settings) {
        super(toolMaterial, attackSpeed, settings);
        this.unlocalizedName = unlocalizedName;
        this.material = material;
    }

    public GSWHoe(GSWMaterial material, ToolMaterial toolMaterial, float attackSpeed, Settings settings) {
        this(material.unlocalizedName() + "_hoe", material, toolMaterial, attackSpeed, settings);
    }

    @Override
    public String unlocalizedName() {
        return unlocalizedName;
    }

    @Override
    public void register() {
        Registry.register(Registry.ITEM, new Identifier(MOD_ID, unlocalizedName), this);
    }

    @Override
    public int getColor(ItemStack stack, int tintIndex) {
        if (tintIndex == 1) {
            return material.getPrimaryColor();
        } else {
            return -1;
        }
    }

    @Override
    public void onClientInit() {
        ColorProviderRegistry.ITEM.register(this, this);
    }

}

File: gsw-mod/src/main/kotlin/io/felipeandrade/gsw/item/tool/GSWHoe.kt
package io.felipeandrade.gsw.item.tool

import io.felipeandrade.gsw.GSWMod.Companion.MOD_ID
import io.felipeandrade.gsw.material.GSWMaterial
import net.minecraft.data.server.recipe.RecipeExporter
import net.minecraft.data.server.recipe.RecipeProvider
import net.minecraft.data.server.recipe.RecipeProvider.getRecipeName
import net.minecraft.data.server.recipe.ShapedRecipeJsonBuilder
import net.minecraft.item.HoeItem
import net.minecraft.item.ItemConvertible
import net.minecraft.item.ToolMaterial
import net.minecraft.recipe.book.RecipeCategory
import net.minecraft.registry.Registries
import net.minecraft.registry.Registry
import net.minecraft.util.Identifier

class GSWHoe(
    val material: GSWMaterial,
    toolMaterial: ToolMaterial,
    attackSpeed: Float,
    settings: Settings,
    attackDamage: Int = 1,
    val unlocalizedName: String = material.unlocalizedName + "_hoe",
) : HoeItem(toolMaterial, attackDamage, attackSpeed, settings), GSWTool {

    override fun register() {
        Registry.register(Registries.ITEM, Identifier(MOD_ID, unlocalizedName), this)
    }

    override fun offerRecipe(exporter: RecipeExporter, ingot: ItemConvertible, handle: ItemConvertible) {
        ShapedRecipeJsonBuilder.create(RecipeCategory.TOOLS, this)
            .pattern("##")
            .pattern(" |")
            .pattern(" |")
            .input('#', ingot)
            .input('|', handle)
            .criterion(RecipeProvider.hasItem(ingot), RecipeProvider.conditionsFromItem(ingot))
            .offerTo(exporter, Identifier(MOD_ID, getRecipeName(this)))
    }
}